### PR TITLE
kotlinx.serialization and Inbyo rock throwing

### PR DESCRIPTION
**Pull request by 0ffz**

### CODE REVIEW

File: src/main/java/com/mineinabyss/mobzy/MobzyCommands.kt [hunk 1]
```diff
@@ -39,7 +42,7 @@ class MobzyCommands : IdofrontCommandExecutor(), TabCompleter {
 
             commandGroup {
                 val entityType by +StringArg("entity type")
-                val radius by +CustomArg<Int>("radius", { arg.toInt() }) {
+                val radius by +IntArg("radius") {
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/MobzyCommands.kt]: i think `Arg<Int>` is cleaner

File: src/main/java/com/mineinabyss/mobzy/MobzyAPI.kt [hunk 2]
```diff
@@ -138,21 +137,23 @@ val EntityTypes<*>.name: String get() = this.f()
...
 val EntityTypes<*>.mobTemplate get() = mobName.toTemplate()
-
+fun String.toTemplate() = templates[toEntityTypeID()] ?: error("Template for $this not found")
 val MobTemplate.type get() = types[name.toEntityTypeID()] ?: error("No entity type found for template $this")
 
 fun Location.spawnEntity(name: String) = spawnEntity(name.toEntityType())
 fun Location.spawnEntity(type: EntityTypes<*>) = mobzy.mobzyTypes.spawnEntity(type, this)
 
+
+fun registerHardCodedTemplate(mob: String, template: MobTemplate): MobTemplate =
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/MobzyAPI.kt]: please add javadocs to this function (and all public functions)

File: src/main/java/com/mineinabyss/mobzy/MobzyType.kt [hunk 3]
```diff
@@ -48,39 +46,33 @@ class MobzyType {
...
+        }
+        return map.toMap()
+    }
+
+    /** Clears all stored [types], and [templates] */
     internal fun reload() { //TODO move all the CustomType related reload stuff here
         _types.clear()
-        _templateNames.clear()
-        _templates.clear()
+        _templates = mapOf()
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/MobzyType.kt]: why is _types mutable but _templates not?

File: src/main/java/com/mineinabyss/mobzy/mobs/MobTemplate.kt [hunk 4]
```diff
@@ -1,81 +1,42 @@
...
-                       var attackDamage: Double? = null,
-                       var minExp: Int? = null,
-                       var maxExp: Int? = minExp,
-                       var isAdult: Boolean = true,
-                       var deathCommands: List<String> = ArrayList(),
-                       var drops: List<MobDrop> = ArrayList()) : ConfigurationSerializable {
+@Serializable
+data class MobTemplate(
+        @SerialName("name") private val _name: String? = null,
+        val model: Int,
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/MobTemplate.kt]: this variable name is worse.
model should refer to the model object

File: src/main/java/com/mineinabyss/mobzy/mobs/MobTemplate.kt [hunk 5]
```diff
@@ -1,81 +1,42 @@
...
+        @SerialName("model-material") val modelMaterial: Material = Material.DIAMOND_SWORD,
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/MobTemplate.kt]: https://google.github.io/styleguide/jsoncstyleguide.xml

File: src/main/java/com/mineinabyss/mobzy/pathfinders/MobzyPathfinderGoal.kt [hunk 6]
```diff
@@ -21,8 +21,13 @@ abstract class MobzyPathfinderGoal(protected val mob: CustomMob, private val coo
...
 
-    protected var lastHit: Long = 0
-    protected val cooledDown get() = lastHit < System.currentTimeMillis() - cooldown
+    private var cooldownStart: Long = 0
+
+    fun restartCooldown() {
+        cooldownStart = System.currentTimeMillis()
+    }
+
+    protected val cooledDown get() = cooldownStart < System.currentTimeMillis() - cooldown
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/pathfinders/MobzyPathfinderGoal.kt]: why is this protected?

File: src/main/java/com/mineinabyss/mobzy/pathfinders/Navigation.kt [hunk 7]
```diff
@@ -1,27 +1,24 @@
...
 
 class Navigation(private val navigationAbstract: NavigationAbstract, private val handle: EntityInsentient) {
-
-    val doneNavigating
-        get() = navigationAbstract.n()
-
-    val pathSearchRange
-        get() = handle.getAttributeInstance(GenericAttributes.FOLLOW_RANGE).value.toFloat()
-
+    val doneNavigating get() = navigationAbstract.m()
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/pathfinders/Navigation.kt]: you should look at bedrock scripting for thoughts on how to structure this code https://bedrock.dev/

File: src/main/java/com/mineinabyss/mobzy/pathfinders/controllers/MZControllerMoveFlying.kt [hunk 8]
```diff
@@ -16,7 +16,7 @@ class MZControllerMoveFlying(private val mob: FlyingMob) : ControllerMove(mob) {
     override fun a() {
         if (h == Operation.MOVE_TO && j-- <= 0) {
             j += Random.nextInt(2, 7)
-            var vec3d = Vec3D(b - mob.x, c - mob.y, d - mob.z)
+            var vec3d = Vec3D(b - mob.locX, c - mob.locY, d - mob.locZ)
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/pathfinders/controllers/MZControllerMoveFlying.kt]: why rename this?

File: src/main/java/com/mineinabyss/mobzy/mobs/CustomMob.kt [hunk 9]
```diff
@@ -30,9 +30,9 @@ interface CustomMob {
     val living: LivingEntity get() = entity.bukkitEntity as LivingEntity
     val template: MobTemplate
     val staticTemplate: MobTemplate get() = entity.entityType.mobTemplate
-    val x: Double get() = living.location.x
-    val y: Double get() = living.location.y
-    val z: Double get() = living.location.z
+    val locX: Double get() = living.location.x
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/CustomMob.kt]: why rename?

File: src/main/java/com/mineinabyss/mobzy/mobs/CustomMob.kt [hunk 10]
```diff
@@ -104,15 +104,16 @@ interface CustomMob {
             if (!entity.world.isClientSide) {
                 if (world.gameRules.getBoolean(GameRules.DO_MOB_LOOT)) {
                     val killer = killer?.bukkitEntity
-                    val looting = killer?.inventory?.itemInMainHand?.enchantments?.get(Enchantment.LOOT_BONUS_MOBS) ?: 0
-                    CraftEventFactory.callEntityDeathEvent(entity, template.chooseDrops(looting))
+                    val heldItem = killer?.inventory?.itemInMainHand
+                    val looting = heldItem?.enchantments?.get(Enchantment.LOOT_BONUS_MOBS) ?: 0
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/CustomMob.kt]: i would put a layer of abstraction here.

File: src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt [hunk 11]
```diff
@@ -1,77 +1,48 @@
...
 import kotlin.random.Random
 
-data class MobDrop(val item: ItemStack,
-                   val minAmount: Int = 1,
-                   val maxAmount: Int = 1,
-                   val dropChance: Double = 1.0) : ConfigurationSerializable {
+@Serializable
+data class MobDrop(
+        val item: SerializableItemStack,
+        @SerialName("cooked-item") val cooked: SerializableItemStack? = null,
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt]: json style

File: src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt [hunk 12]
```diff
@@ -1,77 +1,48 @@
...
+        @SerialName("cook-time") val cookTime: Int = 160,
+        @SerialName("min-amount") val minAmount: Int = 1,
+        @SerialName("max-amount") val maxAmount: Int = 1,
+        @SerialName("drop-chance") val dropChance: Double = 1.0) {
+    init {
+        if (cooked != null) {
+            val cookedItem = cooked.toItemStack()
+            //TODO pretty hacky way of adding recipes. Won't reload them from config if they've changed.
+            try {
+                addFurnaceRecipe(ChatColor.stripColor(cookedItem.itemMeta!!.displayName.color('§').replace(' ', '_'))!!, cookedItem, item.toItemStack(), cookExp, cookTime, mobzy)
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt]: lol why are you adding color to the names?

File: src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt [hunk 13]
```diff
@@ -1,77 +1,48 @@
...
+            }
+        }
+    }
+
     /**
      * @return The amount of items to be dropped, or null if the drop does not succeed
      * TODO I'd like to use exactly what Minecraft's existing system is, but I can't seem to find a way to reuse that.
      */
-    fun chooseDrop(lootingLevel: Int = 0): ItemStack? {
+    fun chooseDrop(lootingLevel: Int = 0, fire: Int = 0): ItemStack? {
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt]: why is fire int vs bool?

File: src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt [hunk 14]
```diff
@@ -1,77 +1,48 @@
...
         val lootingPercent = lootingLevel / 100.0
         val lootingMaxAmount: Int = if (dropChance >= 0.5) (maxAmount + lootingLevel * Random.nextDouble()).roundToInt() else maxAmount
         val lootingDropChance = if (dropChance >= 0.10) dropChance * (10.0 + lootingLevel) / 10.0 else dropChance + lootingPercent
         return if (Random.nextDouble() < lootingDropChance) {
-            val drop = item.clone()
+            val drop = if (fire > 0 && cooked != null) cooked.toItemStack() else item.toItemStack()
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt]: this is very common behavior.
why not abscract?

File: src/main/java/com/mineinabyss/mobzy/pathfinders/flying/PathfinderGoalFlyDamageTarget.kt [hunk 15]
```diff
@@ -10,7 +10,7 @@ class PathfinderGoalFlyDamageTarget(mob: FlyingMob) : MobzyPathfinderGoal(mob) {
     override fun shouldKeepExecuting(): Boolean = false
 
     override fun execute() {
-        lastHit = System.currentTimeMillis()
+        restartCooldown()
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/pathfinders/flying/PathfinderGoalFlyDamageTarget.kt]: nice change

File: src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/ItemThrowable.kt [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package com.mineinabyss.mobzy.pathfinders.hostile
+
+import org.bukkit.entity.LivingEntity
+import org.bukkit.inventory.ItemStack
+
+interface ItemThrowable {
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/ItemThrowable.kt]: i would like to think their are more atributes about how items should be thrown?
also only targeting living entities seems questionable.

File: src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    override fun shouldKeepExecuting() = shouldExecute() && !navigation.doneNavigating
+
+    override fun init() {
+        navigation.moveToEntity(target!!, 1.0)
+    }
+
+    override fun reset() {
+        navigation.stopNavigation()
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt]: what is the base MobzyPathfinderGoal action for this?

File: src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+    override fun execute() {
+        val target = target ?: return
+
+        if (distance < minChaseRad)
+            navigation.stopNavigation()
+
+        if (cooledDown && distance > minShootRad) {
+            restartCooldown()
+            (mob as ItemThrowable).throwItem(target)
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt]: should accuracy be a mob property?

File: src/main/java/com/mineinabyss/mobzy/spawning/SpawnTask.kt [hunk 19]
```diff
@@ -26,14 +26,14 @@ class SpawnTask : BukkitRunnable() {
...
+            Bukkit.getOnlinePlayers().toList().toPlayerGroups().forEach { playerGroup ->
                 val regionManager = container[BukkitAdapter.adapt(playerGroup[0].world)] ?: return@forEach
                 val toSpawn: MutableList<MobSpawnEvent> = mutableListOf()
 
                 //STEP 1: Generate array of ChunkSpawns around player group
-               val spawnChunkGrid = SpawnChunkGrid(playerGroup.map { it.location }, MobzyConfig.minChunkSpawnRad, MobzyConfig.maxChunkSpawnRad)
+                val spawnChunkGrid = SpawnChunkGrid(playerGroup.map { it.location }, MobzyConfig.minChunkSpawnRad, MobzyConfig.maxChunkSpawnRad)
 
-                val customMobs = spawnChunkGrid.allChunks.customMobs
+                val customMobs = spawnChunkGrid.allChunks.filter { it.isLoaded }.customMobs
```
Comment (lizelive) [inline src/main/java/com/mineinabyss/mobzy/spawning/SpawnTask.kt]: makes sense. are you sure it's needed?

**Comment by lizelive** [comment]
general tip. prs should only do one thing

File: build.gradle [hunk 20]
```diff
@@ -112,6 +113,8 @@ shadowJar {
 
     minimize {
         exclude(dependency("org.jetbrains.kotlin:.*"))
+        exclude(dependency("org.jetbrains.kotlinx:kotlinx-serialization-runtime:.*"))
```
[the remark this answers is not in the document]
Comment (0ffz) [reply build.gradle]: Yeah, been meaning to do that and would probably be fitting to go in this PR, i'll do that tomorrow

File: build.gradle [hunk 21]
```diff
@@ -89,18 +89,19 @@ dependencies {
...
-//        implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.14.0" // JVM dependency
+    implementation "com.charleskorn.kaml:kaml:0.16.1"
+    implementation("org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.20.0") // JVM dependency
 
     if (useLocalGuiy.toBoolean()) implementation project(':guiy')
     else implementation 'com.github.mineinabyss:guiy:kotlin-dsl-SNAPSHOT'
 
     if (useLocalIdofront.toBoolean()) implementation project(':idofront')
-    else implementation 'com.mineinabyss:idofront:0.1.+'
+    else implementation 'com.mineinabyss:idofront:0.1.2-alpha.11'
```
[the remark this answers is not in the document]
Comment (0ffz) [reply build.gradle]: The + was using alpha.10 anyways, wonder why it didn't automatically find 11. I'm not actually sure when we are supposed to be using alpha with the current naming scheme, maybe @Derongan could clear things up

File: src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt (hunk 11 above)
Comment (0ffz) [reply src/main/java/com/mineinabyss/mobzy/mobs/MobDrop.kt]: Is json camelCased? I wonder if it makes sense to do that given that I can't find anything specific for yaml, though this is the way spigot serializes items.

File: src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt (hunk 17 above)
Comment (0ffz) [reply src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt]: Does nothing on reset

File: src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt (hunk 18 above)
Comment (0ffz) [reply src/main/java/com/mineinabyss/mobzy/pathfinders/hostile/PathfinderGoalThrowItems.kt]: Will move the trowing method into here probably, and let that be configurabel

File: src/main/java/com/mineinabyss/mobzy/spawning/SpawnTask.kt (hunk 19 above)
Comment (0ffz) [reply src/main/java/com/mineinabyss/mobzy/spawning/SpawnTask.kt]: Timings report showed some lag spikes coming from mobzy related to chunks. I think loading up the unloaded ones synchronously was causing them. Guess we'll see how much it helps soon.

File: src/main/java/com/mineinabyss/mobzy/MobzyAPI.kt [hunk 22]
```diff
@@ -38,17 +38,17 @@ fun debug(message: String, colorChar: Char? = null) {
...
  */
 val Entity.isRenamed
-    get() = if (toNMS().isCustomMob || customName == null) false else customName != this.mobzyID
+    get() = if (toNMS().isCustomMob || customName == null) false else customName != this.mobName
 
 /**
  * The mobzy ID for a registered custom mob.
  */
-val Entity.mobzyID get() = toNMS().mobzyID
+val Entity.mobName get() = toNMS().mobName
```
[the remark this answers is not in the document]
Comment (0ffz) [reply src/main/java/com/mineinabyss/mobzy/MobzyAPI.kt]: Will address in a future PR when I clean up some stuff within the plugin.

**Pull request merged**